File: frontend/public/js/compare.js
document.addEventListener('DOMContentLoaded', function () {
    const compareForm = document.getElementById('compare-form');
    // Updated ID to match template
    const compareBtn = document.getElementById('compare-selected-btn') || document.getElementById('compare-btn');

    if (compareForm && compareBtn) {
        const checkboxes = compareForm.querySelectorAll('.compare-checkbox');
        const updateCompareButton = () => {
            const checkedCheckboxes = compareForm.querySelectorAll('.compare-checkbox:checked');
            const checkedCount = checkedCheckboxes.length;
            
            // Enable button if at least 1 label is selected
            compareBtn.disabled = checkedCount < 1;

            if (checkedCount >= 1) {
                const selectedFormat = checkedCheckboxes[0].getAttribute('data-format');
                checkboxes.forEach(cb => {
                    if (!cb.checked) {
                        // Disable if:
                        // 1. Format mismatch
                        // 2. Already reached 3 selections
                        // 3. Permanent disabled
                        const formatMismatch = cb.getAttribute('data-format') !== selectedFormat;
                        const limitReached = checkedCount >= 3;
                        const permanentDisabled = cb.hasAttribute('data-permanent-disabled');
                        
                        if (formatMismatch || limitReached || permanentDisabled) {
                            cb.disabled = true;
                        } else {
                            cb.disabled = false;
                        }
                    }
                });
            } else {
                // Reset all (except permanently disabled)
                checkboxes.forEach(cb => {
                    if (!cb.hasAttribute('data-permanent-disabled')) cb.disabled = false;
                });
            }
        };

        compareForm.addEventListener('change', (event) => {
            if (event.target.classList.contains('compare-checkbox')) updateCompareButton();
        });

        const showErrorBanner = (msg) => {
            let banner = document.getElementById('error-banner');
            if (!banner) {
                banner = document.createElement('div');
                banner.id = 'error-banner';
                banner.style.cssText = 'background-color: #f8d7da; color: #721c24; padding: 10px; margin-bottom: 15px; border: 1px solid #f5c6cb; border-radius: 4px; text-align: center;';
                // Insert before the table or panels
                const container = compareForm.querySelector('.selection-panels') || compareForm.querySelector('.table-container');
                if (container) {
                    container.parentNode.insertBefore(banner, container);
                } else {
                    compareForm.insertBefore(banner, compareForm.firstChild);
                }
            }
            banner.textContent = msg;
            banner.style.display = 'block';
            banner.scrollIntoView({ behavior: 'smooth', block: 'center' });
            setTimeout(() => { banner.style.display = 'none'; }, 8000);
        };

        compareBtn.addEventListener('click', async (e) => {
            e.preventDefault();
            if (compareBtn.disabled) return;

            const originalText = compareBtn.innerHTML;
            compareBtn.innerHTML = '<span>⏳</span> Loading...';
            compareBtn.disabled = true;

            try {
                const formData = new FormData(compareForm);
                // Need to manually append submit button value if needed, but here we just need fields.
                
                const response = await fetch(compareForm.action, {
                    method: 'POST',
                    body: formData
                });

                if (response.ok) {
                    const html = await response.text();
                    document.open();
                    document.write(html);
                    document.close();
                } else {
                    const data = await response.json();
                    if (data.unsupported_set_id) {
                        // Show banner
                        showErrorBanner(data.error);
                        
                        // Disable the specific item
                        const badCheckbox = compareForm.querySelector(`.compare-checkbox[value="${data.unsupported_set_id}"]`);
                        if (badCheckbox) {
                            badCheckbox.checked = false;
                            badCheckbox.disabled = true;
                            badCheckbox.setAttribute('data-permanent-disabled', 'true'); // Prevent re-enabling
                            
                            // Visual feedback on the row/panel
                            const row = badCheckbox.closest('tr') || badCheckbox.closest('.panel');
                            if (row) {
                                row.style.opacity = '0.5';
                                row.style.pointerEvents = 'none';
                                row.title = "This label has an unsupported format.";
                                
                                // Add a visual indicator text if possible
                                const statusBadge = row.querySelector('.status-badge');
                                if (statusBadge) {
                                    statusBadge.textContent = 'Unsupported';
                                    statusBadge.className = 'status-badge'; // Reset classes
                                    statusBadge.style.backgroundColor = '#6c757d';
                                }
                            }
                        }
                        
                        // Re-evaluate button state
                        updateCompareButton();
                    } else {
                        showErrorBanner(data.error || 'An error occurred.');
                    }
                }
            } catch (err) {
                console.error(err);
                showErrorBanner('Network error occurred. Please try again.');
            } finally {
                // If we are still on the same page (error case)
                if (document.readyState === 'complete' || document.readyState === 'interactive') {
                    compareBtn.innerHTML = originalText;
                    updateCompareButton(); 
                }
            }
        });
        updateCompareButton();
    }

    const compareSectionTitles = document.querySelectorAll('.compare-section-title');
    const expandAllBtn = document.getElementById('expand-all-btn');
    const collapseAllBtn = document.getElementById('collapse-all-btn');

    compareSectionTitles.forEach(title => {
        title.addEventListener('click', () => {
            const section = title.closest('.compare-section');
            section.classList.toggle('expanded');
        });
    });

    if (expandAllBtn && collapseAllBtn) {
        expandAllBtn.addEventListener('click', () => {
            document.querySelectorAll('.compare-section').forEach(section => section.classList.add('expanded'));
        });
        collapseAllBtn.addEventListener('click', () => {
            document.querySelectorAll('.compare-section').forEach(section => section.classList.remove('expanded'));
        });
    }

    const aiSummaryBtn = document.getElementById('ai-summary-btn');
    const aiSummaryLoading = document.getElementById('ai-summary-loading');
    const aiSummaryOutput = document.getElementById('ai-summary-output');
    
    // --- AI Summary Toggle Logic ---
    const toggleAiBtn = document.getElementById('toggle-ai-summary-btn');
    const aiContentWrapper = document.getElementById('ai-summary-content-wrapper');
    const aiHeader = document.querySelector('.ai-summary-header');

    if (toggleAiBtn && aiContentWrapper) {
        const toggleAiSummary = (e) => {
            if (e) e.stopPropagation();
            const isHidden = aiContentWrapper.style.display === 'none';
            aiContentWrapper.style.display = isHidden ? 'block' : 'none';
            toggleAiBtn.innerHTML = isHidden ? '&#8722;' : '&#43;';
            
            // Adjust margin if showing/hiding
            aiHeader.style.marginBottom = isHidden ? '15px' : '0';
        };

        toggleAiBtn.addEventListener('click', toggleAiSummary);
        if (aiHeader) aiHeader.addEventListener('click', toggleAiSummary);
    }

    if (aiSummaryBtn && aiSummaryLoading && aiSummaryOutput && typeof comparisonData !== 'undefined' && typeof selectedLabelsMetadata !== 'undefined') {
        
        // --- Helper to fetch summary ---
        async function fetchSummary(forceRefresh = false, generateIfMissing = true) {
            aiSummaryOutput.innerHTML = '';
            aiSummaryLoading.style.display = 'block';
            aiSummaryBtn.disabled = true;

            const differingSections = [];
            comparisonData.forEach(section => {
                if (!section.is_same) {
                    const content1 = section.contents[0] ? stripHtmlTags(section.contents[0]) : '';
                    const content2 = section.contents[1] ? stripHtmlTags(section.contents[1]) : '';
                    differingSections.push({ title: section.title, content1: content1, content2: content2 });
                }
            });

            const label1Name = selectedLabelsMetadata[0] ? selectedLabelsMetadata[0].brand_name : 'Label 1';
            const label2Name = selectedLabelsMetadata[1] ? selectedLabelsMetadata[1].brand_name : 'Label 2';

            try {
                const response = await fetch('/api/dashboard/ai_compare_summary', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({
                        set_ids: typeof currentSetIds !== 'undefined' ? currentSetIds : [],
                        differing_sections: differingSections,
                        label1_name: label1Name,
                        label2_name: label2Name,
                        force_refresh: forceRefresh,
                        generate_if_missing: generateIfMissing
                    }),
                });

                const data = await response.json().catch(() => ({}));

                if (!response.ok) {
                    if (generateIfMissing) { // Only show error if we expected a result
                        aiSummaryOutput.innerHTML = `<p style="color: red;">Error: ${data.error || 'Network response was not ok.'}</p>`;
                    }
                } else if (data.summary) {
                    // Clean up markdown code blocks if AI wrapped the HTML
                    let cleanSummary = data.summary.replace(/```html/g, '').replace(/```/g, '');
                    
                    if (typeof marked !== 'undefined' && marked.parse) {
                        aiSummaryOutput.innerHTML = marked.parse(cleanSummary);
                    } else {
                        console.warn('Marked library not loaded. Falling back to plain text.');
                        aiSummaryOutput.innerHTML = cleanSummary.replace(/\n/g, '<br>');
                    }
                    
                    aiSummaryBtn.innerHTML = '<span>&#x21bb;</span> Re-generate Summary';
                } else if (!generateIfMissing) {
                    // No cache found, do nothing
                } else if (data.error) {
                    aiSummaryOutput.innerHTML = `<p style="color: red;">Error: ${data.error}</p>`;
                } else {
                    aiSummaryOutput.innerHTML = `<p>Could not generate summary.</p>`;
                }

            } catch (error) {
                console.error('Error during AI summary fetch:', error);
                if (generateIfMissing) {
                    aiSummaryOutput.innerHTML = `<p style="color: red;">An error occurred while generating the AI summary.</p>`;
                }
            } finally {
                aiSummaryLoading.style.display = 'none';
                aiSummaryBtn.disabled = false;
            }
        }

        // --- Init Check for Saved Summary ---
        fetchSummary(false, false); // Check only

        // --- Click Handler ---
        aiSummaryBtn.addEventListener('click', async () => {
            // If text is "Re-generate", we force refresh. 
            // Otherwise (first time), we just fetch (generateIfMissing=true).
            const isRegen = aiSummaryBtn.textContent.includes('Re-generate');
            fetchSummary(isRegen, true);
        });
    }
});

window.exportSectionToMarkdown = function(event, index) {
    if (event) event.stopPropagation(); 
    
    const rootSection = comparisonData[index];
    const rootKey = rootSection.key ? rootSection.key.toString() : '';
    const labels = selectedLabelsMetadata;
    
    let md = `# Comparison: ${rootSection.title}\n\n`;
    const dateStr = new Date().toISOString().split('T')[0];
    md += `**Date:** ${dateStr}\n\n`;
    
    labels.forEach((l, i) => {
        md += `**Label ${i+1}:** ${l.brand_name} (${l.manufacturer_name || 'N/A'})\n`;
        md += `- Set ID: ${l.set_id}\n`;
        md += `- Date: ${l.effective_time || 'N/A'}\n\n`;
    });
    
    md += '---\n\n';
    
    // Collect root section and any descendants
    // Descendants are subsequent sections whose key starts with "rootKey." (PLR) or "rootKey " (Non-PLR normalized)
    // Since comparisonData is flat sorted, descendants should immediately follow.
    
    let sectionsToExport = [rootSection];
    
    for (let i = index + 1; i < comparisonData.length; i++) {
        const nextSec = comparisonData[i];
        if (!nextSec.key) break;
        
        const nextKey = nextSec.key.toString();
        
        // PLR Logic: "5" -> "5.1", "5.2"
        const isPlrDescendant = nextKey.startsWith(rootKey + '.');
        
        // Non-PLR Logic: "WARNINGS" -> "WARNINGS GENERAL" (via normalized keys with space)
        // Ensure rootKey is not empty to avoid matching everything
        const isNonPlrDescendant = rootKey.length > 0 && nextKey.startsWith(rootKey + ' ');
        
        if (isPlrDescendant || isNonPlrDescendant) {
            sectionsToExport.push(nextSec);
        } else {
            // Stop if sequence breaks (assuming sorted list keeps families together)
            // But wait, "5" -> "5.1" -> "5.2". All good.
            // "5" -> "6". Stops.
            break;
        }
    }

    sectionsToExport.forEach(section => {
        md += `### ${section.title}\n\n`;
        
        section.contents.forEach((content, i) => {
            const labelName = labels[i] ? labels[i].brand_name : `Label ${i+1}`;
            md += `#### ${labelName}\n\n`;
            
            if (!content) {
                md += '*(Section not found or empty)*\n\n';
            } else {
                // Clean HTML for Markdown
                let text = content.replace(/<br\s*\/?>/gi, '\n');
                text = text.replace(/<\/p>/gi, '\n\n');
                text = text.replace(/<\/div>/gi, '\n\n');
                text = text.replace(/<\/li>/gi, '\n');
                text = text.replace(/<li>/gi, '- ');
                
                // Remove remaining tags using DOMParser
                var doc = new DOMParser().parseFromString(text, 'text/html');
                text = doc.body.textContent || '';
                
                // Fix multiple newlines
                text = text.replace(/\n\s*\n/g, '\n\n').trim();
                
                md += `${text}\n\n`;
            }
        });
        md += '---\n\n';
    });
    
    // Add AI Prompt
    md += '## AI Analysis Prompt\n\n';
    md += `> **Instruction:** Please analyze the content provided above for the section "${rootSection.title}" (and any included subsections) from the different drug labels.\n`;
    md += '> \n';
    md += '> **Goal:** Compare the information and generate a summary that emphasizes:\n';
    md += '> 1. **Commonalities:** What information is consistent across all labels?\n';
    md += '> 2. **Differences:** What are the key differences (e.g., dosing, warnings, populations)?\n';
    md += '> 3. **Implications:** Briefly highlight any clinical significance of these differences.\n';
    
    // Download
    const blob = new Blob([md], { type: 'text/markdown' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    // Use root section title for filename
    const safeTitle = rootSection.title.replace(/[^a-z0-9]/gi, '_').substring(0, 50); 
    a.download = `Comparison_${safeTitle}.md`;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
};

window.exportFullComparisonJson = function() {
    const labels = selectedLabelsMetadata;
    const comparisonStructure = [];

    // Helper to clean HTML
    const cleanHtml = (html) => {
        if (!html) return null;
        let text = html.replace(/<br\s*\/?>/gi, '\n');
        text = text.replace(/<\/p>/gi, '\n\n');
        text = text.replace(/<\/div>/gi, '\n\n');
        text = text.replace(/<\/li>/gi, '\n');
        text = text.replace(/<li>/gi, '- ');
        var doc = new DOMParser().parseFromString(text, 'text/html');
        return (doc.body.textContent || '').replace(/\n\s*\n/g, '\n\n').trim();
    };

    comparisonData.forEach(section => {
        const entry = {
            title: section.title,
            key: section.key,
            contents: []
        };

        section.contents.forEach((content, i) => {
            const labelId = labels[i] ? labels[i].set_id : `label_${i}`;
            entry.contents.push({
                set_id: labelId,
                text: cleanHtml(content),
                is_missing: content === null
            });
        });

        comparisonStructure.push(entry);
    });

    const exportObject = {
        metadata: {
            export_date: new Date().toISOString(),
            comparison_title: typeof comparisonTitle !== 'undefined' ? comparisonTitle : 'Drug Label Comparison',
            labels: labels
        },
        comparison_data: comparisonStructure,
        system_prompt: `You are an expert regulatory affairs specialist and clinical pharmacist.
Your task is to analyze the provided drug label comparison data.
The data contains a list of sections. For each section, text from multiple drug labels is provided.

Please perform the following for the entire comparison:
1. Identify the most significant clinical differences between the labels, focusing on Indications, Warnings, and Dosage.
2. Summarize the commonalities.
3. If there are discrepancies in contraindications or boxed warnings, highlight them immediately.
4. Provide a high-level executive summary of how these labels differ in terms of safety and efficacy profiles.`
    };

    const blob = new Blob([JSON.stringify(exportObject, null, 2)], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `Full_Comparison_${new Date().toISOString().slice(0,10)}.json`;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
};

File: backend/labelcomp/compare.py
import json
import hashlib
from datetime import datetime
from database import db, ComparisonSummary
from dashboard.services.ai_handler import call_llm

def get_comparison_summary(user, set_ids, comparison_data, label_names, force_refresh=False):
    """
    Fetches a cached comparison summary or generates a new one.
    """
    if not set_ids:
        return None
    
    # 1. Check Cache
    sorted_ids = sorted(set_ids)
    ids_str = json.dumps(sorted_ids)
    ids_hash = hashlib.sha256(ids_str.encode('utf-8')).hexdigest()
    
    if not force_refresh:
        cached = ComparisonSummary.query.filter_by(set_ids_hash=ids_hash).first()
        if cached:
            return cached.summary_content

    # 2. Prepare data for AI
    differing_sections = []
    total_chars = 0
    MAX_CHARS_FOR_AI = 100000 # 100k chars limit to prevent timeout

    for section in comparison_data:
        if 'contents' in section:
            if section.get('is_same') is True or section.get('is_empty') is True:
                continue
            contents = [c or 'N/A' for c in section.get('contents', [])]
            differing_sections.append({
                'title': section.get('title'),
                'contents': contents
            })
            total_chars += sum(len(c) for c in contents)

        # Backward compatibility for content1/content2
        elif 'content1' in section and 'content2' in section:
            c1 = section.get('content1') or 'N/A'
            c2 = section.get('content2') or 'N/A'
            differing_sections.append({
                'title': section.get('title'),
                'contents': [c1, c2]
            })
            total_chars += len(c1) + len(c2)
        # Or the full format from backend (contents list)
        elif not section.get('is_same') and not section.get('is_empty'):
            contents = section.get('contents', [])
            if contents:
                c_list = [c or 'N/A' for c in contents]
                differing_sections.append({
                    'title': section.get('title'),
                    'contents': c_list
                })
                total_chars += sum(len(c) for c in c_list)

    if not differing_sections:
        return "<p>No substantive differences were identified between the labels for comparison.</p>"

    # 3. Handle data volume limit
    if total_chars > MAX_CHARS_FOR_AI:
        return f"""
            <div style="padding: 1.5rem; background-color: #fff7ed; border: 1px solid #fed7aa; border-radius: 12px; color: #9a3412;">
                <h3 style="margin-top: 0; color: #9a3412;">Comparison Data Too Large</h3>
                <p>These labeling documents contain a very large amount of differing content ({total_chars:,} characters), which exceeds our automated reasoning limit.</p>
                <p>To analyze these differences, please use the <strong>Export</strong> button below to download the comparison data as a JSON file, which you can then upload to <strong>ELSA</strong> or other specialized analysis tools.</p>
            </div>
        """

    # 4. Generate via AI
    try:
        summary_content = summarize_comparison_logic(user, differing_sections, label_names)
    except Exception as e:
        print(f"AI Summary Error: {str(e)}")
        return f"""
            <div style="padding: 1.5rem; background-color: #fef2f2; border: 1px solid #fee2e2; border-radius: 12px; color: #991b1b;">
                <h3 style="margin-top: 0; color: #991b1b;">AI Analysis Interrupted</h3>
                <p>We encountered an issue while generating the clinical summary (likely a timeout due to complex content).</p>
                <p>Please use the <strong>Export</strong> button to download the raw comparison data for review in <strong>ELSA</strong>.</p>
            </div>
        """
    
    # 5. Update Cache
    existing = ComparisonSummary.query.filter_by(set_ids_hash=ids_hash).first()
    if existing:
        existing.summary_content = summary_content
        existing.timestamp = datetime.utcnow()
    else:
        new_summary = ComparisonSummary(
            set_ids_hash=ids_hash,
            set_ids=ids_str,
            summary_content=summary_content
        )
        db.session.add(new_summary)
    
    db.session.commit()
    return summary_content

def summarize_comparison_logic(user, differing_sections, label_names):
    """
    The actual AI prompt and call logic. Supports multiple labels.
    """
    summary_parts = []
    for section in differing_sections:
        title = section.get('title', 'Unknown Section')
        contents = section.get('contents', [])
        
        summary_parts.append(f"--- Section: {title} ---\n")
        for idx, content in enumerate(contents):
            label_name = label_names[idx] if idx < len(label_names) else f"Label {idx+1}"
            summary_parts.append(f"{label_name}:\n{content}\n\n")

    combined_diff_text = "".join(summary_parts)
    
    label_list_str = ", ".join([f"'{n}'" for n in label_names])
    num_labels = len(label_names)

    system_prompt = f"""
        You are an expert AI analyst for the FDA. Your sole function is to identify and summarize the key substantive differences between {num_labels} drug labeling documents.

        **Core Task:**
        1.  Analyze the provided text, which contains the content of {num_labels} different drug labels: {label_list_str}.
        2.  Identify the most critical differences, focusing on safety, efficacy, indications, contraindications, and warnings.
        3.  Generate a concise "Overall Critical Differences" executive summary.
        4.  Generate a section-by-section summary of notable differences.

        **CRITICAL OUTPUT FORMATTING RULES:**
        -   Your response MUST be ONLY raw HTML.
        -   DO NOT include any preamble, explanation, conversational text, or markdown code blocks (```html). Your response must start directly with the <h3> tag.

        The entire output must follow this exact structure:
        <h3>Overall Critical Differences</h3>
        <ul>
            <li>A summary of the most important difference between the versions.</li>
            <li>Another key difference summary.</li>
        </ul>

        <div class="summary-section">
            <h4>[Section Name, e.g., Indications and Usage]</h4>
            <ul>
                <li>Detail of a difference found in this section across the labels.</li>
            </ul>
        </div>

        If a section has no significant differences, DO NOT include a heading for it.
        If there are no differences at all, output only: <p>No substantive differences were identified between the labels.</p> 
    """
    
    user_message = (
        f"Compare the drug labels for {label_list_str}. "
        f"Generate the HTML summary based on the content below.\n\n"
        f"--- DOCUMENT CONTENT ---\n{combined_diff_text}"
    )
    
    return call_llm(user, system_prompt, user_message)
